model/src/application/interaction/application_command/data/mod.rs

```diff
@@ -26,50 +34,186 @@ pub struct CommandData {
 
 /// Data received when a user fills in a command option.
 ///
-/// Note: user, channel, role, and mentionable option types will be returned as
-/// a [`String`] option here.
-///
 /// Refer to [the discord docs] for more information.
 ///
-/// [the discord docs]: https://discord.com/developers/docs/interactions/application-commands#interaction-applicationcommandinteractiondataoption
-#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
+/// [the discord docs]: https://discord.com/developers/docs/interactions/application-commands#application-command-object-application-command-interaction-data-option-structure
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct CommandDataOption {
+    pub name: String,
+    pub value: CommandOptionValue,
+}
+
+/// Value of a [`CommandDataOption`].
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum CommandOptionValue {
+    String(String),
+    Integer(i64),
+    Boolean(bool),
+    User(UserId),
+    Channel(ChannelId),
+    Role(RoleId),
+    Mentionable(GenericId),
+    SubCommand(Vec<CommandDataOption>),
+    SubCommandGroup(Vec<CommandDataOption>),
+    Number(Number),
+}
+
+#[derive(Debug, Deserialize)]
+struct CommandDataOptionRaw<'a> {
+    name: String,
+    #[serde(rename = "type")]
+    kind: CommandOptionType,
+    value: Option<CommandOptionValueRaw<'a>>,
+    #[serde(default)]
+    options: Option<Vec<CommandDataOption>>,
+}
+
+#[derive(Debug, Deserialize, Serialize)]
 #[serde(untagged)]
-pub enum CommandDataOption {
-    String {
-        name: String,
-        value: String,
-    },
-    Integer {
-        name: String,
-        value: i64,
-    },
-    Boolean {
-        name: String,
-        value: bool,
-    },
-    SubCommand {
-        name: String,
-        #[serde(default)]
-        options: Vec<CommandDataOption>,
-    },
+enum CommandOptionValueRaw<'a> {
+    String(Cow<'a, str>),
```

Comment:
When is `Cow<str>` an optimization over `String`?

Reply:
I think it will save a allocation when serializing in this case, is that the idea @LeSeulArtichaut?

Reply:
I suppose so, could it be replaced by `&'a str` maybe? And maybe have `CommandDataOptionRaw` contain `&'a str` & `Option<&'a [CommandDataOption]>`. I don't know Serde well enough to say if it makes a difference.

Reply:
Correct, this is to avoid allocating a temporary string when serializing. I probably slapped a `Cow<str>` because I don't know serde well myself and couldn't find out if using a `&str` works when deserializing, which I don't a big deal because we'll allocate to put it in the `CommandOptionValue` afterwards anyway.

Reply:
We can reevaluate it at some other time